azurerm_web_application_firewall_policy - support managed_rules

sirlatrom opened the pull request:
Adds `managed_rules` to `azurerm_web_application_firewall_policy`. Until this commit, at least since SDK version 2019-09-01 (as imported in the touched file), it was not possible to create a new resource of that type, as reported in #5354.

~Note: Also streamline singular plurality for optional blocks allowing multiple repetitions to comply with how other `azurerm_` resources use singular for block names when there may be multiple specified.~

Fixes #5727.
Fixes #5354.

sirlatrom commented on azurerm/internal/services/network/resource_arm_web_application_firewall_policy.go:
```diff
@@ -402,12 +577,102 @@ func flattenArmWebApplicationFirewallPolicyPolicySettings(input *network.PolicyS
 	result := make(map[string]interface{})
 
-	result["enabled"] = input.State == network.WebApplicationFirewallEnabledStateDisabled
+	result["enabled"] = input.State == network.WebApplicationFirewallEnabledStateEnabled
```
This one snuck in, and fixes a bug where `policy_settings.enabled` is set to the opposite of the actual value.

sirlatrom commented:
Ping @katbyte @tombuildsstuff, this is a prerequisite for #6105 since the test in that one requires successfully creating an `azurerm_web_application_firewall_policy`.

katbyte commented on azurerm/internal/services/network/resource_arm_web_application_firewall_policy.go:
```diff
@@ -143,6 +143,87 @@ func resourceArmWebApplicationFirewallPolicy() *schema.Resource {
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"rule_set_type": {
```
I think we can just use type as rule set is implied?
```suggestion
									"type": {
```

sirlatrom replied:
Done in 9fa3f6ea0.

katbyte commented on azurerm/internal/services/network/resource_arm_web_application_firewall_policy.go:
```diff
@@ -143,6 +143,87 @@ func resourceArmWebApplicationFirewallPolicy() *schema.Resource {
+										Required:     true,
+										ValidateFunc: validation.NoZeroValues,
+									},
+									"rule_set_version": {
```
```suggestion
									"version": {
```

sirlatrom replied:
Done in 9fa3f6ea0.

katbyte commented on azurerm/internal/services/network/resource_arm_web_application_firewall_policy.go:
```diff
@@ -332,11 +418,102 @@ func expandArmWebApplicationFirewallPolicyPolicySettings(input []interface{}) *n
+	exclusions := v["exclusion"].([]interface{})
+	managedRuleSets := v["managed_rule_set"].([]interface{})
+
+	result := network.ManagedRulesDefinition{
```
We could just return the struct?
```suggestion
	return &network.ManagedRulesDefinition{
```

sirlatrom replied:
Done in 9fa3f6ea0.

katbyte commented on website/docs/r/web_application_firewall_policy.html.markdown:
```diff
@@ -130,6 +163,42 @@ The `policy_setting` block supports the following:
+
+* `selector` - (Optional) Describes field of the matchVariable collection.
+
+* `selector_match_operator` - (Required) Describes operator to be matched.
```
COuld we list all possible values here?

sirlatrom replied:
Done in 9fa3f6ea0.

katbyte commented on website/docs/r/web_application_firewall_policy.html.markdown:
```diff
@@ -130,6 +163,42 @@ The `policy_setting` block supports the following:
+
+The `exclusion` block supports the following:
+
+* `match_variable` - (Required) The name of the Match Variable.
```
COuld we list all possible values here?

sirlatrom replied:
Done in 9fa3f6ea0.

sirlatrom commented on azurerm/internal/services/network/resource_arm_web_application_firewall_policy.go:
```diff
@@ -48,7 +48,7 @@ func resourceArmWebApplicationFirewallPolicy() *schema.Resource {
 			"resource_group_name": azure.SchemaResourceGroupNameDiffSuppress(),
 
-			"custom_rules": {
+			"custom_rule": {
```
There can be no existing users of the resource as it does not work, as described in #5354. I thought that before making the resource actually work, this would be a good chance to align naming with other resources. The existing docs even mention a `custom_rule` block:

![image](https://user-images.githubusercontent.com/425633/77225636-fd653900-6b70-11ea-8c3b-15dd7193d56c.png)

The same goes for `match_condition` and `match_variable`.

A single block with a plural name would indicate that the block configures all such elements, e.g. custom rules, whereas multiple blocks with a singular name indicate that they each configure such an element.

sirlatrom replied:
I see now that the `ManagedRules` field was introduced in #5004 (2e69260f3a) between dependency `github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-07-01/network` and `github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-09-01/network`, but the `azurerm_web_application_firewall_policy` resource was not updated to reflect this new required field.

This of course means that there will be existing users of this resource that have used the `custom_rules` name for the property before the `ManagedRules` field was introduced, so I suppose I'll revert this change after all, and update the docs to all use the actual property names instead of the singular forms.

sirlatrom replied:
Fixed in 7aeb41b5f.

sirlatrom commented:
> Thanks for making those changes @sirlatrom. I just noticed there is not test for the new properties, could we set them in the complete test? once that's done this should be good to merge.

This should be done in cefd1c8.

sirlatrom commented:
> 
> 
> @sirlatrom,
> 
> we can't add a new required property, one it clearly works without it 🙂 and it would break the resource for existing users. As can be see from the tests:
> 
> ```
> Test Failed
> 
> ------- Stdout: -------
> === RUN   TestAccAzureRMWebApplicationFirewallPolicy_basic
> === PAUSE TestAccAzureRMWebApplicationFirewallPolicy_basic
> === CONT  TestAccAzureRMWebApplicationFirewallPolicy_basic
> --- FAIL: TestAccAzureRMWebApplicationFirewallPolicy_basic (0.76s)
>     testing.go:640: Step 0 error: config is invalid: "managed_rules": required field is not set
> ```

The property is required on the ARM side since API version 2019-08-01, but was not implemented when this resource was switched to that version. I'll try and add a reasonable default value.

sirlatrom commented:
@katbyte When providing an empty ManagedRules object I get this API error from Azure, so I think it really must be specified:

> ```
> Error: Error creating Web Application Firewall Policy "foobarPolicy" (Resource Group "rgFooBar"): network.WebApplicationFirewallPoliciesClient#CreateOrUpdate: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="ApplicationGatewayFirewallManagedRuleSetsNoValidCrsRuleSetsAttached" Message="Firewall policy /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/rgFooBar/providers/Microsoft.Network/ApplicationGatewayWebApplicationFirewallPolicies/foobarPolicy does not have any valid Core Rule Set attached to it." Details=[]
> ```

the essence of which is: **foobarPolicy does not have any valid Core Rule Set attached to it**.

It's all caused by #5004 which updated the SDK version from 2019-07-01 to 2019-09-01 without implementing the required field. See here: https://github.com/terraform-providers/terraform-provider-azurerm/blame/91d54d8ef85c82b2aa38723a12f5709dd9ef1abe/azurerm/resource_arm_web_application_firewall_policy.go#L8.

Anyone with a provider version prior to this change will not be affected. Anyone with the current version and later are currently not able to create firewall policy resources, and will need this fix to be able to do so again. Given how both the SDK and the API say the field is required and cannot just be created without properly configuring it, I would claim it actually _does_ have to be required.

If we have to fill out default values for the ManagedRules field, it has to contain a choice of a managed rule set, which I don't think either of us are able to choose without breaking other things for the end users. Hence, I believe they should choose one themselves when upgrading to the provider version that adds this field.

katbyte commented:
@sirlatrom, thanks for investigating this. I agree with everything you said and looks like we just have to mark it required and move forward as checking in our CI systems i can see the error and that this has been broken for quite some time. Triggering CI and taking a look now

sirlatrom commented:
> @sirlatrom, this looks good aside for we need to update the existing tests so they pass, basic with the minimal rules, complete with everything set.

@katbyte Thank you for your patience! I've added the required properties to the tests in 2f15e45, and corrected the `managed_rules.managed_rule_set`'s `Required` to `true` as it is indeed required. I've also adjusted #6105 accordingly. I had also made a typo when referring to the `managed_rule_set` property in the tests.

---

One last thing I saw was that in the `waf_configuration` block, the `rule_set_type` has a default value of `OWASP` and the `rule_set_version` and `disabled_rule_group.rule_group_name` properties have `StringInSlice` validation functions. The default type is `OWASP` and the validation functions include different versions and rule group names across the possible types.

I'm not sure if those defaults and lists of valid values will always be up to date, so while I could add the same ones to the equivalent properties in the `managed_rules.managed_rule_set.rule_group_override.rule_group_name`/`disabled_rules` properties; but I would both have to duplicate those lists of valid values, resulting in double maintenance and the risk of drift, and additionally, those lists may grow out of date resulting in the potential need of the provider needing a new release _just_ to support a new rule set type or version or disabled rule.

WDYT?

sirlatrom commented:
@katbyte Looks like I have more work to do then. I imagine the tests fail because I forgot to access `managed_rules` as a list, and `policy_settings` I'm not sure about. I'll find time to look at it tomorrow CEST.

sirlatrom commented:
@katbyte OK, it should be done in 6c30ebdbb now, but I don't have access to run acceptance tests in my current environment.

sirlatrom commented:
PS: 6731c99 includes the validation functions you suggested in https://github.com/terraform-providers/terraform-provider-azurerm/pull/6126#pullrequestreview-385989469.

sirlatrom commented:
@katbyte I've tried creating a policy on our subscription, and it comes back with these values when running `az network application-gateway waf-policy show ...`:

```json
...
  "policySettings": {
    "fileUploadLimitInMb": 100,
    "maxRequestBodySizeInKb": 128,
    "mode": "Prevention",
    "requestBodyCheck": true,
    "state": "Enabled"
  },
...
```

So I'm able to set the fields using the patched provider. When I run plan the next time, I'm told there are no changes to be made, so that should indicate the settings are read properly, too. We might be dealing with the defaults, here, which are set to `"mode": "Detection"` and `"state": "Enabled"`.

sirlatrom commented:
@katbyte One option would be to add the `policy_settings` block to the basic config in the test, even if it is not required.

sirlatrom commented:
The test that fails looks unrelated.

sirlatrom commented:
@katbyte Tests pass after rebase on origin/master which includes #6540. Please let me know if acceptance tests still fail after I added `policy_settings` to the basic test.
